### ISSUE TITLE
Porting to .net core

### ISSUE THREAD

**Issue by wmccullough** [canton7/RestEase#11]
Hey there,

I spent a few hours today porting this to .net core. The team I work with love RestEase. If we contributed back, would you consider taking the port?

I still have a ways to go with Moq not being ported yet, but I'm looking to contribute to them too in order to get it working with this project.

I figured I'd reach out before I assumed anything. I was also weary of starting a pull request only because I'm not sure of the best way to organize the project files considering that netfull and netcore would be living side by side.

In a nutshell, wanting to help, but I don't want to step on your toes!

**Comment by canton7** [comment]
Awesome stuff! Sure, I'd definitely be open to supporting more platforms.

I'm halfway through an attempt to build the interface implementations at compile-time, rather than runtime, which should open up the door to supporting lots of other platforms with ease (we'll be able to release as a PCL, for instance). Maybe that would help you?

I've got Monday free, so I'll hopefully be able to spend some time actually getting stuck into .NET Core, and making sure I understand what sort of effort adding support would take.

**Comment by canton7** [comment]
Right, I've been doing some digging.

The .NET Core reflection model is 1) currently a bit different, and 2) [about to change anyway](https://blogs.msdn.microsoft.com/dotnet/2016/05/27/making-it-easier-to-port-to-net-core/). Since building the interface implementations relies on an awful lot of reflection, I'm not sure how keen I am to rush out and port this to .NET Core, if everything's going to move back in a few months anyway.

The other approach, and the once I've been considering for creating a PCL which supports targets like Xamarin, is to do the implementation creation at compile-time, rather than runtime, by inserting a custom build task into the user's csproj. However .NET Core has its `project.json` format, and I can't find much on whether it supports `UsingTask`, and whether NuGet supports adding custom build tasks automatically, both of which would be required. That said, `project.json` is going to be dropped anyway, but I'm not sure what the timescale for this is: I get the impression that it's VS "15" only?

So, the issues are whether to wait for the .NET reflection API to stabilise and continue generating implementations at runtime (I'll have to do work to generate them at compile-time for targets like Xamarin, but that will be independent), potentially resulting in two PCLs for different targets, or wait until .NET Core moves back to msbuild and use the compile-time generating for everything.

Do you have any input on this? It sounds like you're more embedded in the .NET Core world than I am...

**Comment by wmccullough** [comment]
After a few of the revelations came out about the future of .net core, I'm not sure this is a good idea either. The way that they made it sound, core is going to end up going to RTM, but then change to something else as part of some global .net strategy. I'm not sure it's too smart to even invest time in core until it really stabilizes.

**Comment by canton7** [comment]
Yeah, that's kind of my worry. 

In short, I want to support .NET Core in principle, and if someone else puts the effort in and adds support in a way that doesn't disrupt things too much (e.g., don't just make a copy of the project and modify bits, as that will turn into a maintainability nightmare), then I'll happily merge it. However I probably won't put any effort in myself until it becomes clear that .NET Core is actually vaguely stable.

**Comment by wmccullough** [comment]
I can completely understand that. I now have a fear that two months after RTM, core will disappear completely and become a strategy for how they want to change .net going forward vs a standalone product. I'm going to wait as well. 

In the mean time, is there anything in the .net full version that I can help out with? I love this framework.

**Comment by canton7** [comment]
Not right now, but thanks for the offer. Unless you've got any other suggestions, I think the next big thing is PCL support. I've been putting that off, but this Monday's a public holiday so I'm hoping to get a bit done then. After that, I may ask you to test stuff and tell me where I've broken everything :smile:

**Comment by wmccullough** [comment]
I'm totally down to help test! I have tons of enterprise grade projects that I can test this against!

**Comment by canton7** [comment]
Having seen [the Source Generator feature for VS "15"](https://github.com/dotnet/roslyn/blob/master/docs/features/generators.md), I'm tempted to say that this is a better approach than a custom build task...

**Comment by wmccullough** [comment]
Hey @canton7! Now that the dust has settled (as well as some of the drama), I was thinking about picking this back up again. I can start looking into the source generator or any other task you're interested in if you'd like?

**Comment by canton7** [comment]
Hey! There's actually an open PR for adding core support -  I've been away all week and haven't had the chance to look at it yet.

**Comment by canton7** [comment]
Would you mind giving version 1.0.1.116 a go? It's available on this custom NuGet feed: https://ci.appveyor.com/nuget/restease.

It should support netstandard1.3, net40, and net45.

**Comment by wmccullough** [comment]
@canton7 I'm trying it this afternoon! Sorry, been swamped lately.

**Comment by wmccullough** [comment]
@canton7 Gah, my environment at work is locked down so that I can't hit the nuget feed on appveyor. I'll have to try it at home.

**Comment by wmccullough** [comment]
@canton7 Success! Works flawlessly, tested most scenarios too!

**Comment by canton7** [comment]
Awesome, thanks for testing!

I need to tweak the README, as I need to move from gitlink back to symbolsource, but I'll hopefully get it released to nuget soon.

**Comment by canton7** [comment]
Released as version 1.1.1. Thanks for your help!

**Comment by fazouane-marouane** [comment]
That's good news! Glad I could be of help 😄 
Btw, Xamarin is [now supporting netstandard](https://blog.xamarin.com/net-standard-library-support-for-xamarin/).
